### ISSUE TITLE
WebSocket ends prematurely in chrome, results in server erroring on writes

### ISSUE THREAD

**Issue by ThisIsMissEm** [websockets/ws#2148]
### Is there an existing issue for this?

- [X] I've searched for any related issues and avoided creating a duplicate issue.

### Description

This is related to the mastodon streaming server, essentially we're seeing some activity that suddenly sends a spike of like 100k messages to a websocket connection, individually, and the socket ends up "stalling" in chrome (that was the connection state in chrome developer tools). We added a log line recently to give more information about the errors (as for some reason a large stacktrace is being captured), and the error we're getting is:

```
Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
```

We do have an explicit check before the `ws.send` that is `if (ws.readyState !== ws.OPEN) {`, but that doesn't seem to help.

We listen on the following websocket events:
- error
- close
- message
- pong

And listen to the `connection` event from the WebSocket Server.

Is there something else we should be listening for to ensure we don't attempt to send to a closed socket?

### ws version

^8.12.1

### Node.js Version

v16.20.0 (but we've observed this behaviour on more recent versions)

### System

multiple, I can reproduce on MacOS latest with Chrome latest, but we see this in production with unix-based servers (like ubuntu, debian, and arch)

**Comment by ThisIsMissEm** [comment]
The full error stacktrace is:

```
Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
     at new NodeError (node:internal/errors:387:5)
     at errorBuffer (node:internal/streams/writable:519:14)
     at afterWrite (node:internal/streams/writable:503:5)
     at onwrite (node:internal/streams/writable:479:7)
     at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:87:12)
```

And the constructor is just "Error"

**Comment by lpinca** [comment]
Those are probably buffered writes that can't go through. There is not much to do apart from checking the [`websocket.bufferedAmount`](https://github.com/websockets/ws/blob/master/doc/ws.md#websocketbufferedamount) and stop writing if it grows too much.

**Comment by lpinca** [comment]
Anyway the error comes from Node.js core and a minimal test case to reproduce the issue would be useful.

**Comment by lpinca** [comment]
Also are you sure the error is coming from the `WebSocket` object? We don't reemit socket errors. You can get socket errors in the `websocket.send()` callback but that is not specified in the issue description. Finally you can get `EPIPE` and `ECONNRESET` errors but it shouldn't be possible to get a "Cannot call write after a stream was destroyed" error.

**Comment by ThisIsMissEm** [comment]
I enabled `NODE_DEBUG=stream` and saw a `STREAM 18901: onWriteComplete -89 undefined` which tracing that error back to the darwin posix error codes indicates a `EQFULL` error meaning `Interface output queue is full`, which I totally would expect, given I'm trying to flush 200k messages to a websocket near instantly

**Comment by ThisIsMissEm** [comment]
Yes, we're getting this in the `ws.send` callback, and we do sometimes get EPIPE, but mostly just the above. Logging `bufferedAmount` by the time my load tool finishes pushing events through redis (which get passed to `ws.send`) the `ws.bufferedAmount` is 13150676.

I'm logging `ws.bufferedAmount` each `ws.send`, it stays at 0 for a while, then jumps to these numbers:

```
ERR! BUFFERED: 0 
ERR! BUFFERED: 0 
ERR! BUFFERED: 77 
ERR! BUFFERED: 154 
ERR! BUFFERED: 231 
ERR! BUFFERED: 308 
ERR! BUFFERED: 385 
ERR! BUFFERED: 462 
ERR! BUFFERED: 539 
ERR! BUFFERED: 616 
ERR! BUFFERED: 693 
ERR! BUFFERED: 770 
ERR! BUFFERED: 847 
ERR! BUFFERED: 924 
```

And just keeps growing

**Comment by lpinca** [comment]
How big is each message? A possible way to see if the issue is reproducible without `ws` would be to use a dummy WebSocket server created via `http.createServer()` and send frames created via `WebSocket.frame()` to the socket obtained via the `'upgrade'` event.

**Comment by ThisIsMissEm** [comment]
This is the load test I'm using.. it is excessive:

```
const redis = require('redis');

async function main() {
  const client = redis.createClient({
    url: "redis://localhost:6379",
  });

  client.on('error', (err) => log.error('Redis Client Error!', err));
  await client.connect();


  for(let i = 0; i < 200_000; ++i) {
    client.publish('timeline:public', "{\"event\":\"delete\",\"payload\":\"110492033014171175\"}");
    client.publish('timeline:public:local', "{\"event\":\"delete\",\"payload\":\"110492033014171175\"}");
  }

  await client.quit()
}

main();
```

Essentially the websocket is subscribed to the redis pubsub for `'timeline:public'`, and tries to send the message out over websocket instantly.

The reason the load test is so epically high is because we've a scenario where a certain schedule job triggers many hundreds of thousands of redis publish calls.

**Comment by lpinca** [comment]
Ok, but if the behavior is reproducible without using `ws`, for example using a simple `net.Socket`, I think there is not much we can do.

**Comment by ThisIsMissEm** [comment]
My current theory is just that we're producing WAY too many messages all at once, and that the load we're putting on websocket connections is simply so beyond what is reasonable that yes, errors should be expected.

**Comment by lpinca** [comment]
We want the callback of `websocket.send()` to be called only once and if the first write fails, the second should also fails. Plus the socket is corked so it should happen as a single write.

**Comment by ThisIsMissEm** [comment]
> We want the callback of `websocket.send()` to be called only once and if the first write fails, the second should also fails. Plus the socket is corked so it should happen as a single write.

Ah, okay, so we're not gonna get a scenario of the first failing and the second succeeding?

**Comment by ThisIsMissEm** [comment]
okay, cool. I'm gonna close this as a "won't fix", as I am aware that the sheer load in terms of messages in a given timeframe is so ridiculously high that it just shouldn't be expected that a client would handle that (chrome really dislikes it)

**Comment by lpinca** [comment]
Anyway if something like this

```js
const net = require('net');

const server = net.createServer();

server.on('connection', function (socket) {
  for (let i = 0; i < 200000; i++) {
    socket.write(
      '{"event":"delete","payload":"110492033014171175"}',
      function (err) {
        console.error(err);
      }
    );

    socket.on('error', function () {});
  }
});

server.listen(8080, '127.0.0.1', function () {
  console.log('Sever listening on 127.0.0.1:8080')
});
```

produces the error described here, I think we should report it to Node.js core as it is misleading.

**Comment by lpinca** [comment]
This is what I was talking about in https://github.com/websockets/ws/issues/2148#issuecomment-1591795481.

```js
const assert = require('assert');
const crypto = require('crypto');
const http = require('http');
const WebSocket = require('ws');

const GUID = '258EAFA5-E914-47DA-95CA-C5AB0DC85B11';

const data = `<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
  </head>
  <body>
    <script>
      (function () {
        var ws = new WebSocket('ws://localhost:8080');

        ws.onopen = function () {
          console.log('open');
          ws.send('ready');
        };
      })();
    </script>
  </body>
</html>`;

const server = http.createServer();

server.on('request', function (req, res) {
  res.setHeader('Content-Type', 'text/html');
  res.end(data);
});

server.on('upgrade', function (req, socket) {
  const key = crypto
    .createHash('sha1')
    .update(req.headers['sec-websocket-key'] + GUID)
    .digest('base64');

  socket.on('error', console.error);
  socket.once('data', (chunk) => {
    assert.strictEqual(chunk.length, 11);
    assert.deepStrictEqual(chunk.slice(0, 2), Buffer.from([0x81, 0x85]));

    socket.resume();

    const data = JSON.stringify({
      event: 'delete',
      payload: '110492033014171175'
    });
    const frame = Buffer.concat(
      WebSocket.Sender.frame(data, { fin: true, opcode: 1 })
    );

    for (let i = 0; i < 200000; i++) {
      socket.write(frame, function (err) {
        if (err) console.error(err);
      });
    }

    socket.write(Buffer.from([0x88, 0x00]));
  });

  socket.write(
    [
      'HTTP/1.1 101 Switching Protocols',
      'Upgrade: websocket',
      'Connection: Upgrade',
      `Sec-WebSocket-Accept: ${key}`,
      '\r\n'
    ].join('\r\n')
  );
});

server.listen(8080, function () {
  console.log('Server listening on *:8080');
});
```

It freezes Chrome but I don't get the error.

**Comment by lpinca** [comment]
> It freezes Chrome but I don't get the error.

There is another side effect though. On my machine (macOS) the socket does not emit the `'close'` event even is I close the browser and the socket is kept open indefinitely. It must be destroyed manually.

**Comment by ThisIsMissEm** [comment]
> > It freezes Chrome but I don't get the error.
> 
> There is another side effect though. On my machine (macOS) the socket does not emit the `'close'` event even is I close the browser and the socket is kept open indefinitely. It must be destroyed manually.

Yeah, that too! The ping/pong does eventually call ws.terminate() but not fast enough to avoid errors. I think we'll just need to have a high watermark on bufferedAmount (is this bytes or messages?) and when it exceeds that amount, just drop messages — it's not ideal but it'll do
